route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -0,0 +1,176 @@
+        auto shortest_path = wm_->getRoute()->shortestPath();
+        double current_progress = wm_->routeTrackPos(current_loc).downtrack;
+        double speed_progress = current_speed_;
+        double total_maneuver_length = current_progress + mvr_duration_ * RouteFollowingPlugin::TWENTY_FIVE_MPH_IN_MS;
+        int last_lanelet_index = findLaneletIndexFromPath(current_lanelet.second.id(), shortest_path);
+        if(last_lanelet_index == -1)
+        {
+            ROS_WARN_STREAM("Cannot find current lanelet in shortest path!");
+            return true;
+        }
+        while(current_progress < total_maneuver_length && last_lanelet_index < shortest_path.size())
+        {
+            double end_dist = wm_->routeTrackPos(shortest_path[last_lanelet_index].centerline2d().back()).downtrack;
+            double dist_diff = end_dist - current_progress;
+
+            resp.new_plan.maneuvers.push_back(composeManeuverMessage(current_progress, end_dist, speed_progress, shortest_path[last_lanelet_index].id(), ros::Time::now()));
+
+            current_progress += dist_diff;
+            speed_progress = RouteFollowingPlugin::TWENTY_FIVE_MPH_IN_MS;
+            if(current_progress >= total_maneuver_length || last_lanelet_index == shortest_path.size() - 1)
+            {
+                break;
+            }
+
+            auto following_lanelets = wm_->getRoute()->followingRelations(shortest_path[last_lanelet_index]);
+            if(following_lanelets.size() == 0)
+            {
+                ROS_WARN_STREAM("Cannot find the following lanelet.");
+                return true;
+            }
+            if(identifyLaneChange(following_lanelets, shortest_path[last_lanelet_index + 1].id()))
+            {
+                ++last_lanelet_index;
+            }
+            else
+            {
+                ROS_WARN_STREAM("Cannot find the next lanelet in the current lanelet's successor list!");
+                return true;
+            }
+        }
+        if(resp.new_plan.maneuvers.size() == 0)
+        {
+            ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
+        }
+        return true;
+    }
+
+    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+    }
+
+    void RouteFollowingPlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        current_speed_ = msg->twist.linear.x;
+    }
+
+    int RouteFollowingPlugin::findLaneletIndexFromPath(int target_id, lanelet::routing::LaneletPath& path)
+    {
+        for(size_t i = 0; i < path.size(); ++i)
+        {
+            if(path[i].id() == target_id)
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    cav_msgs::Maneuver RouteFollowingPlugin::composeManeuverMessage(double current_dist, double end_dist, double current_speed, int lane_id, ros::Time current_time)
```

Comment:
Assuming we can eventually get the speed limits (or other target speed) from the World Model, you'd need to parameterize this method for target speed too, correct? Might be a better idea to add target speed as a parameter here, and then if we do need to use the placeholder `RouteFollowingPlugin::TWENTY_FIVE_MPH_IN_MS` value for now we can do that wherever we'd be invoking this method.

Reply:
Sure. Make sense to me.